### PR TITLE
modifications to extract filesystem for kbuild

### PR DESCRIPTION
**Pull request by priyawadhwa**
Extract symlinks from the tarball, and also add support for a whitelist (files that would be ignored during extraction)

### CODE REVIEW

File: pkg/util/cloud_prepper.go [hunk 1]
```diff
@@ -59,7 +59,7 @@ func (p CloudPrepper) GetFileSystem() (string, error) {
 		return "", err
 	}
 
-	return path, getFileSystemFromReference(ref, p.ImageSource, path)
+	return path, GetFileSystemFromReference(ref, p.ImageSource, path, []string{})
```
Comment (dlorenc) [inline pkg/util/cloud_prepper.go]: You should be able to put nil here.

File: pkg/util/docker_utils.go [hunk 2]
```diff
@@ -115,7 +115,7 @@ func unpackDockerSave(tarPath string, target string) error {
 	}
 
 	for _, layer := range layers {
-		if err = UnTar(bytes.NewReader(layerMap[layer]), target); err != nil {
+		if err = UnTar(bytes.NewReader(layerMap[layer]), target, []string{}); err != nil {
```
Comment (dlorenc) [inline pkg/util/docker_utils.go]: And here.

File: pkg/util/tar_utils.go [hunk 3]
```diff
@@ -65,7 +67,7 @@ func unpackTar(tr *tar.Reader, path string) error {
 				if err := os.MkdirAll(target, mode); err != nil {
 					return err
 				}
-			} else {
+				// In some cases, MkdirAll doesn't change the permissions, so run Chmod
```
Comment (dlorenc) [inline pkg/util/tar_utils.go]: It's usually umask related, but I suppose if the dirs already exist it wouldn't change the permissions either

File: pkg/util/tar_utils.go [hunk 4]
```diff
@@ -106,21 +107,52 @@ func unpackTar(tr *tar.Reader, path string) error {
...
+	return filepath.Walk(p, func(path string, info os.FileInfo, err error) error {
+		if e := os.Chmod(path, 0777); e != nil {
+			logrus.Errorf("Error updating file permissions on %s before removing for symlink creation", path)
+			return e
+		}
+		return os.RemoveAll(path)
+	})
+}
+
+func checkWhitelist(target string, whitelist []string) bool {
```
Comment (dlorenc) [inline pkg/util/tar_utils.go]: I think you want something more precise than prefix matching. Something like /foo would block /foobar when it should only block /foo/bar

File: pkg/util/fs_utils.go [hunk 5]
```diff
@@ -160,3 +160,22 @@ func CheckSameFile(f1name, f2name string) (bool, error) {
 	}
 	return true, nil
 }
+
+// HasFilepathPrefix checks if the given file path begins with prefix
+func HasFilepathPrefix(path, prefix string) bool {
+	path = filepath.Clean(path)
+	prefix = filepath.Clean(prefix)
+	pathArray := strings.Split(path, "/")
```
Comment (dlorenc) [inline pkg/util/fs_utils.go]: Can you use filepath.Split here?

File: pkg/util/tar_utils.go [hunk 6]
```diff
@@ -49,13 +47,16 @@ func unpackTar(tr *tar.Reader, path string) error {
...
-			if err = os.RemoveAll(newName); err != nil {
+			if err = walkAndRemove(newName); err != nil {
 				logrus.Error(err)
 			}
 			continue
 		}
-
 		target := filepath.Join(path, header.Name)
+		// Make sure the target isn't part of the whitelist
+		if checkWhitelist(target, whitelist) {
```
Comment (dlorenc) [inline pkg/util/tar_utils.go]: A debug log here will come in handy later.

File: pkg/util/tar_utils.go [hunk 7]
```diff
@@ -106,21 +106,52 @@ func unpackTar(tr *tar.Reader, path string) error {
...
+				logrus.Errorf("Failed to create symlink between %s and %s: %s", header.Linkname, target, err)
+			}
+		}
 	}
 	return nil
 }
 
+func walkAndRemove(p string) error {
+	return filepath.Walk(p, func(path string, info os.FileInfo, err error) error {
+		if e := os.Chmod(path, 0777); e != nil {
```
Comment (dlorenc) [inline pkg/util/tar_utils.go]: Why do we need to chmod before removal?

Comment (priyawadhwa) [reply pkg/util/tar_utils.go]: Ah sorry, I was trying something out and meant to remove that

File: pkg/util/fs_utils.go (hunk 5 above)
Comment (priyawadhwa) [reply pkg/util/fs_utils.go]: I think it only [splits](https://golang.org/pkg/path/filepath/#Split) along the last separator

File: pkg/util/tar_utils.go (hunk 6 above)
Comment (priyawadhwa) [reply pkg/util/tar_utils.go]: I added it in the checkWhitelist function

File: pkg/util/fs_utils.go (hunk 5 above)
Comment (dlorenc) [reply pkg/util/fs_utils.go]: D'oh. Right.

**Pull request merged**